### ISSUE TITLE
test: Race condition susceptibility in object_store tests

### ISSUE THREAD

**Issue by brandonsov** [influxdata/influxdb_iox#551]
__Steps to reproduce:__
List the minimal actions needed to reproduce the behaviour.

1. Create at least 1 test that adds an entry in any object store (more tests improve the chances)
2. Run object store tests until you see the test failure
3. Confirm you see tests that use `put_get_delete_list` helper function failing

__Expected behaviour:__
All object_store tests pass regardless of order or threads used

__Actual behaviour:__
Tests using `put_get_delete_list` helper function fail due to asserts that require the object store to be in a specific state

__Environment info:__

* Branch: `main`
* Folder: `object_store`
* Command used: `cargo test`
* System info: Linux 5.4.0-56-generic x86_64

__Config:__
Environment file copied with a valid environment variable SERVICE_ACCOUNT credentials file and a bucket that exists for the GCS_BUCKET_NAME environment variable

__Logs:__

In my test below, I've added 6 different unit tests adding data to the gcs object store and removing it from the gcs object store. I suspect all other object store tests are susceptible to this, but I have not confirmed.
```sh
failures:

---- tests::google_cloud_storage::gcs_test stdout ----
thread 'tests::google_cloud_storage::gcs_test' panicked at 'assertion failed: `(left == right)`
  left: `["12345", "123456", "test_file"]`,
 right: `["test_file"]`', object_store/src/lib.rs:838:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    tests::google_cloud_storage::gcs_test

test result: FAILED. 14 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out

```
 

__Notes:__

> I believe the put_get_delete_list test helper function is susceptible to race conditions. Given the rust unit tests are run asynchronously by default, I believe this will frequently throw false negatives if there are other tests that modify the same bucket. I can write up a new issue or modify within this PR if desired.

Ref: https://github.com/influxdata/influxdb_iox/pull/536

Ideas: 
- Don't assert state of object store other than searching the entity added exists 
- Use a global object store reference wrapped in a mutex

**Comment by carols10cents** [comment]
While I agree that the object store tests looking at shared state are racy as you've pointed out, it's actually worse than that... the cloud-storage-rs crate has a global client that can't be used from multiple tests running in parallel: https://github.com/ThouCheese/cloud-storage-rs/issues/60

**Comment by jacobmarble** [comment]
@carols10cents is this concern still valid?

**Comment by carols10cents** [comment]
Yes, the original concern is still valid, but probably low importance.
